onnx/defs/schema.cc

```diff
@@ -478,36 +438,48 @@ OpSchema& OpSchema::FillUsing(std::function<void(OpSchema&)> populator) {
   return *this;
 }
 
-int OpSchema::CalculateOutput(int num_input) const {
-  if (min_output_ == max_output_) {
-    return min_output_;
-  } else if (calculate_output_) {
-    return calculate_output_(num_input);
-  } else {
-    return kCannotComputeNumOutputs;
-  }
-}
-
 void OpSchema::Finalize() {
 #define ENFORCE(x)                                                          \
   do {                                                                      \
     if (!(x))                                                               \
       throw std::logic_error(                                               \
           "ONNX Schema " + name_ + ": failed validating the check: " + #x); \
   } while (0)
-  ENFORCE(min_input_ <= max_input_);
-  ENFORCE(min_output_ <= max_output_);
-  ENFORCE(inputs_.size() >= min_input_);
-  ENFORCE(outputs_.size() >= min_output_);
-  ENFORCE(inputs_.size() <= max_input_);
-  ENFORCE(outputs_.size() <= max_output_);
-  // if max limit is finite - all names should be specified
-  if (max_input_ < std::numeric_limits<int>::max()) {
-    ENFORCE(inputs_.size() == max_input_);
+
+  // Calculate min/max number of inputs.
+  for (int i = 0; i < (int)(inputs_.size()); ++i) {
+    switch (inputs_[i].GetOption()) {
+    case OpSchema::Single:
+      ++min_input_;
+      ++max_input_;
+      break;
+    case OpSchema::Optional:
```

Comment:
this is the case only if the input appears in the end. If (for some reason) we have optional inputs in the middle, I wouldn't count them to the total

Reply:
Not quite get your opinion here please. Are you saying that,
1. We should only allow optional inputs at the end please?
2. If there's optional input defined in the middle (there're Single inputs afterwards), we'll just not use it? in this case, the operator registration is invalid?
Thank you!

Reply:
I would keep the semantic of "optional" same for both input and output. This means, allowing "optional" input and output in the middle with empty string specified.

Reply:
It's fine to allow them in the middle, I'm just saying that reporting min_input in this case is invalid. Imagine you have

0: single
1: optional
2: single

We should either disallow this case or make min_input=max_input=3 for it. Current code min_input=2 and it doesn't make sense. Shall we disallow it for now?